src/transformers/modeling_auto.py

```diff
@@ -286,32 +290,36 @@ def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
-        elif "transfo-xl" in pretrained_model_name_or_path:
-            return TransfoXLModel.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
-        elif "xlnet" in pretrained_model_name_or_path:
-            return XLNetModel.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
-        elif "xlm" in pretrained_model_name_or_path:
-            return XLMModel.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
-        elif "ctrl" in pretrained_model_name_or_path:
-            return CTRLModel.from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
+        config = kwargs.pop("config", None)
+        if not isinstance(config, PretrainedConfig):
+            config = AutoConfig.from_pretrained(pretrained_model_name_or_path, **kwargs)
+
+        if isinstance(config, T5Config):
+            return T5Model.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, DistilBertConfig):
+            return DistilBertModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, AlbertConfig):
+            return AlbertModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, CamembertConfig):
+            return CamembertModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, XLMRobertaConfig):
+            return XLMRobertaModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, RobertaConfig):
+            return RobertaModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, BertConfig):
+            return BertModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, OpenAIGPTConfig):
+            return OpenAIGPTModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, GPT2Config):
+            return GPT2Model.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, TransfoXLConfig):
+            return TransfoXLModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, XLNetConfig):
+            return XLNetModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, XLMConfig):
+            return XLMModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
+        elif isinstance(config, CTRLConfig):
+            return CTRLModel.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
```

Comment:
Wouldn't it be better to do like you did for the configuration, using an`OrderedDict` `CONFIG_MAPPING` so that it is more future-proof?